### ISSUE TITLE
More flexible batching for graph-level properties

### ISSUE THREAD

**Issue by Linux-cpp-lisp** [rusty1s/pytorch_geometric#2030]
## 🚀 Feature

It would be great if `Data` subclasses had more control over the batching behavior of properties that are not per-node. In particular, two features would be very useful:

1. The ability to concatenate some properties along a *new* batch dimension:
```python
class MyData(Data):
    def __cat_dim__(self, key, value):
        if key == 'some_prop':
            return None
        else:
            return super().__cat_dim__(key, value)

# ...
N_data = 3
datas = [MyData(some_prop=torch.randn(4, 7)) for _ in range(N_data)]
batch = Batch.from_data_list(datas)
batch.some_prop.shape # => [N_data, 4, 7]
```
I think `None` is a reasonable choice for this value since it gives a new axis in numpy indexing.

2. The ability to keep, but exempt from batching, properties of the `Data` objects that are conceptually "batch-level" and can be taken once from any one of the individual `Data` objects in the batch:
```python
class MyData(Data):
    def __cat_dim__(self, key, value):
        if key == 'some_prop':
            return False
        else:
            return super().__cat_dim__(key, value)

# ...
N_data = 3
datas = [MyData(some_prop=torch.randn(4, 7)) for _ in range(N_data)]
batch = Batch.from_data_list(datas)
batch.some_prop.shape # => [4, 7]
torch.all(batch.some_prop == datas[-1].some_prop) # => True
```
`False` might make sense here in the sense that it is saying to not concatenate, but there may be a better way of doing this.

Instead of explicitly using `None` and `False`, there could also be constants `Batch.CAT_DIM_NEW` and `Batch.CAT_DIM_DISABLE`, or something similar.

## Motivation

For 1., if I have a per-graph target property that is not scalar (for example, a fixed-size [M, N] Tensor), it would be logically correct for this property to have dimension [M,N] in an individual `Data` object, but to have dimension [n_batch, M, N] in a `Batch` assembled from those `Data`s. Currently, to get this behavior, the property must be stored as [1, M, N] in the individual example and have `__cat_dim__` of 0, which while workable is conceptually incorrect.

For 2., as I said in #2028, 

> In some cases, Data objects may contain per-graph fields that are known to be the same across the entire batch/dataset. In cases where the model code is aware of this and knows how to deal with a missing batch dimension for those fields, it would be nice to avoid constructing an unneeded full [n_batch, ...] set of copies.

In some such scenarios, the value of the property at any *one* of the individual `Data` objects should be taken and stored in the final `Batch`. This can be done by a custom `collate` function, but it would be nice if  `torch_geometric` supported it directly.

**Comment by Linux-cpp-lisp** [comment]
For 2., another option could be allowing a `Data` subclass to provide a list of "batch-level" properties outside of `__cat_dim__` that will be treated separately, and perhaps optionally checked for consistency (checked that they are all equal across `Data`s in the batch, raising an error otherwise).

**Comment by rusty1s** [comment]
I think those are some great suggestions. Are you interested in contributing those?
1. I think `None` is a absolutely fine.
2. Not sure if this is really necessary. I would expect global attributes to be in `dataset.global_attribute` instead of in each individual `data` object.

**Comment by Linux-cpp-lisp** [comment]
Sure!

My thinking for the second case was based around properties that may sometimes be per-graph, or sometimes per-dataset, depending on the dataset in question. In that case it seemed easier to me to have the information inside the Batch regardless so that the training loop doesn't need access to the Dataset object. It's easy when loading individual Data objects to do this in an efficient way, since they can each carry a reference to the single Tensor with the data, but this becomes suddenly memory inefficient when batching happens.

Still, I agree that this may be to specialized for inclusion in the library and may be better implemented by subclassing Batch and wrapping from_data_list. What do you think?

**Comment by Linux-cpp-lisp** [comment]
@rusty1s I've filed a draft pull request implementing the first feature along with a simple unit test. The test suite passes but it would probably be good if you could take a look and see if it changes the API in any problematic ways

**Comment by Linux-cpp-lisp** [comment]
Closed by #2036
